File: ReverseLinkedList/I/ReverseLinkedList.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
	# iteratively 
    def reverseList_1(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None or head.next == None:
        	return head
        pre = None
        cur = head
        while cur != None:
        	tmp = cur.next
        	cur.next = pre
        	pre = cur
        	cur = tmp
        return pre

    # recursively
    def reverseList_2(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if head == None or head.next == None:
        	return head
        newHead = self.reverseList_2(head.next)
        head.next.next = head
        head.next = None
        return newHead

File: ReverseLinkedList/I/ReverseLinkedList.cpp
/**
* Definition for singly-linked list.
*/
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};
class Solution {
public:
	// iteratively 
	ListNode* reverseList_1(ListNode* head) {
		if (head == nullptr)
			return head;
		ListNode* pre = nullptr;
		ListNode* cur = head;
		while (cur != nullptr) {
			ListNode* tmp = cur->next;
			cur->next = pre;
			pre = cur;
			cur = tmp;
		}
		return pre;
	}
	// recursively
	ListNode* reverseList_2(ListNode* head) {
		if (head == nullptr || head->next == nullptr)
			return head;
		ListNode* newHead = reverseList_2(head->next);
		head->next->next = head;
		head->next = nullptr;
		return newHead;
	}
};